x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -245,14 +271,49 @@ private void executeSubPlan(
                 LocalRelation resultWrapper = resultToPlan(tuple.logical, result);
 
                 // replace the original logical plan with the backing result
-                final PhysicalPlan newPlan = plan.transformUp(FragmentExec.class, f -> {
-                    LogicalPlan frag = f.fragment();
-                    return f.withFragment(
-                        frag.transformUp(
-                            InlineJoin.class,
-                            ij -> ij.right() == tuple.logical ? InlineJoin.inlineData(ij, resultWrapper) : ij
-                        )
-                    );
+                final PhysicalPlan newPlan = plan.transformUp(PhysicalPlan.class, p -> {
+                    if (p instanceof FragmentExec f) {
+                        LogicalPlan frag = f.fragment();
+
+                        return f.withFragment(frag.transformUp(LogicalPlan.class, bp -> {
+                            if (bp instanceof InlineJoin ij && ij.right() == tuple.logical) {
+                                return InlineJoin.inlineData(ij, resultWrapper);
+                            } else if (bp instanceof Merge mr && mr.subPlans().stream().anyMatch(sp -> sp == tuple.logical)) {
+                                List<LogicalPlan> newSubPlans = new ArrayList<>(mr.subPlans());
+                                for (int i = 0; i < newSubPlans.size(); i++) {
+                                    if (newSubPlans.get(i) == tuple.logical) {
+                                        newSubPlans.set(i, resultWrapper);
+                                    }
+                                }
+                                return new Merge(mr.source(), newSubPlans);
```

Comment:
If the subplans would have. been children, replacing them would have been possible through node.transform()